pkg/server/credtemplate/builder_test.go

```diff
@@ -0,0 +1,1188 @@
+		X509SVIDTTL:     credtemplate.DefaultX509SVIDTTL,
+		JWTSVIDTTL:      credtemplate.DefaultJWTSVIDTTL,
+		JWTIssuer:       "",
+		AgentSVIDTTL:    credtemplate.DefaultX509SVIDTTL,
+	}, config)
+}
+
+func TestNewBuilderAllowsConfigOverrides(t *testing.T) {
+	configIn := credtemplate.Config{
+		TrustDomain:     td,
+		X509CASubject:   pkix.Name{CommonName: "X509CA"},
+		X509SVIDSubject: pkix.Name{CommonName: "X509SVID"},
+		X509CATTL:       1 * time.Minute,
+		X509SVIDTTL:     2 * time.Minute,
+		JWTSVIDTTL:      3 * time.Minute,
+		JWTIssuer:       "ISSUER",
+		AgentSVIDTTL:    4 * time.Minute,
+	}
+	builder, err := credtemplate.NewBuilder(configIn)
+	require.NoError(t, err)
+
+	configOut := builder.Config()
+
+	// Assert that the Clock and NewSerialNumber are not nil and then set them
+	// to nil before comparing the whole config. Checking the whole config in a
+	// single equality check is more future proof but the defaults for these
+	// fields are hard to compare.
+	assert.NotNil(t, configOut.Clock)
+	configOut.Clock = nil
+	assert.NotNil(t, configOut.NewSerialNumber)
+	configOut.NewSerialNumber = nil
+
+	assert.Equal(t, configIn, configOut)
+}
+
+func TestBuildSelfSignedX509CATemplate(t *testing.T) {
+	for _, tc := range []struct {
+		desc             string
+		overrideConfig   func(config *credtemplate.Config)
+		overrideParams   func(params *credtemplate.SelfSignedX509CAParams)
+		overrideExpected func(expected *x509.Certificate)
+		expectErr        string
+	}{
+		{
+			desc: "defaults",
+		},
+		{
+			desc: "fail to get serial number",
+			overrideConfig: func(config *credtemplate.Config) {
+				config.NewSerialNumber = failNewSerialNumber
+			},
+			expectErr: "failed to get new serial number: oh no",
+		},
+		{
+			desc: "invalid public key",
+			overrideParams: func(params *credtemplate.SelfSignedX509CAParams) {
+				params.PublicKey = nil
+			},
+			expectErr: "x509: unsupported public key type: <nil>",
+		},
+		{
+			desc: "override X509CATTL",
+			overrideConfig: func(config *credtemplate.Config) {
+				config.X509CATTL = time.Minute * 23
+			},
+			overrideExpected: func(expected *x509.Certificate) {
+				expected.NotAfter = now.Add(time.Minute * 23)
+			},
+		},
+		{
+			desc: "override X509CASubject",
+			overrideConfig: func(config *credtemplate.Config) {
+				config.X509CASubject = pkix.Name{CommonName: "OVERRIDE"}
+			},
+			overrideExpected: func(expected *x509.Certificate) {
+				expected.Subject = pkix.Name{CommonName: "OVERRIDE"}
+			},
+		},
+		{
+			desc: "single composer",
+			overrideConfig: func(config *credtemplate.Config) {
+				config.CredentialComposers = []credentialcomposer.CredentialComposer{fakeCC{id: 1}}
+			},
+			overrideExpected: func(expected *x509.Certificate) {
+				expected.Subject.CommonName = "OVERRIDE-1"
+				expected.PolicyIdentifiers = []asn1.ObjectIdentifier{{1}}
+				expected.ExtraExtensions = []pkix.Extension{{Id: makeOID(1), Value: []byte{1}}}
+			},
+		},
+		{
+			desc: "two composers",
```

Comment:
may we set something on the first composer, to be sure it survives from the second one? actual test case is basically overriding all, so we are not sure if the first one ran. (the same comment is for all another tests that do something similar)

Reply:
Let me see what I can do...